File: Logic_API/src/main/java/model/DataModel.java
package model;

import dataTransferObjects.*;
import dataTransferObjectsDB.*;
import sockets.socketHandler;

import javax.jws.soap.SOAPBinding;
import java.util.ArrayList;
import java.util.Arrays;


public class DataModel {
    private socketHandler socket;

    public DataModel() {
        socket = new socketHandler();
    }

    public User addUser(User user) {
        User newUser = null;
        String[] usernames = getAllUsernames();
        for(int i = 0; i < usernames.length; i++)
        {
            if(usernames[i].equals(user.getUsername()))
            {
                return null;
            }
        }

        if (user != null) {
            System.out.println("Model received new user: " + user.getFirstName() + " (" + user.getUsername() + ")");
            String hashedPass = EncryptPassword.getHashedPassword(user.getPassword());
            System.out.println("Users password is hashed from \"" + user.getPassword() +"\" to \"" + hashedPass + "\"");
            if(hashedPass.equals("NoSuchAlgorithmException") || hashedPass.equals("InvalidKeySpecException") || hashedPass.equals("IllegalArgumentException"))
                return null;
            user.setPassword(hashedPass);
            newUser = socket.addNewUser(user);
        }
        return newUser;
    }

    public User getUserObject(String username, String password) {
        try {
            User user = socket.searchUser(username);
            if(user.getPassword().equals(password) || EncryptPassword.autenthicate(password, user.getPassword()))
                return user;
        }
        catch (NullPointerException e)
        {
            System.out.printf("User not found");
        }
        return null;
    }

    public Project[] getUserProject(String username) {
        ProjectDB[] projectsDB = socket.getProjects(username);
        if (projectsDB != null) {
            Project[] projects = new Project[projectsDB.length];
            for (int i = 0; i < projectsDB.length; i++)
                projects[i] = projectsDB[i].convertToProject();
            return projects;
        } else
            return new Project[0];
    }

    public Project createUserProject(String username, String pName) {
        String[] projectNames = getAllProjectNames();
        for(String s: projectNames)
        {
            if(s.equals(pName))
            {
                return null;
            }
        }
        ProjectDB projectDB = socket.addNewProject(username, pName);
        Project project = projectDB.convertToProject();
        System.out.println("New Project added: " + project.getName());
        return project;
    }

    public User inviteUserProject(String username, String pName) {
        String[] usernames = getAllUsernames();
        for(int i = 0; i < usernames.length; i++)
        {
            if(usernames[i].equals(username) && isUserInProject(username, pName))
            {
                User user = socket.inviteUserToProject(username, pName);
                System.out.println("New user added to project " + pName + ": " + user.getUsername());
                return user;
            }
        }
        return null;
    }

    private boolean isUserInProject(String username, String pName)    {
        WorkGroup workGroup = socket.getWorkgroup(pName, "getWorkgroup");
        ArrayList<User> developers = workGroup.getAllDevelopers();
        // This is working for Kiril
        if((workGroup.getCreator() != null && workGroup.getCreator().getUsername().equals(username)) ||
                (workGroup.getScrumMaster() != null && workGroup.getScrumMaster().getUsername().equals(username)) ||
                (workGroup.getProductOwner() != null && workGroup.getProductOwner().getUsername().equals(username)))
            return false;

        for (User u: developers
             ) {
            if (u.getUsername().equals(username))
                return false;
        }

        return true;
    }

    public User deleteUserInvitation(String username, String pName) {
        User user = socket.deleteUserInvitation(username, pName);
        System.out.println("Invitation is deleted from project " + pName + ": " + user.getUsername());
        return user;
    }

    public String[] getInvitationsByUsername(String username) {
        String[] invitations = socket.getInvitations(username);
        System.out.println("Invitations received...");
        return invitations;
    }

    public Project getAcceptedInvitation(String username, String pName) {
        System.out.println("We are in DataModelClass");
        Project project = socket.getAcceptedInvitation(username, pName);
        System.out.println("User: " + username + " has accepted the " + pName + " project' invitation!");
        return project;
    }

    public Project setRole(String username, String pName, String pRole) {
        Project project = new Project();
        WorkGroup workGroup = socket.getWorkgroup(pName, "getWorkgroup");
        if (pRole.equals("SCRUM Master")) {


            if (workGroup.getScrumMaster() == null) {

                project = socket.setRole(username, pName, "setScrumMaster");
            } else {
                project = socket.setRole(username, pName, "updateScrumMaster");
            }


            System.out.println("User: " + username + " has the new role: " + pRole + " in project: " + pName);
            return project;
        } else if (pRole.equals("Product Owner")) {

            if (workGroup.getProductOwner() == null) {
                project = socket.setRole(username, pName, "setProductOwner");
            } else {
                project = socket.setRole(username, pName, "updateProductOwner");
            }

            System.out.println("User: " + username + " has the new role: " + pRole + " in project: " + pName);
            return project;
        }

        return null;
    }

    public Project createBacklogItem(String projectName, BackLogItem backLogItem) {
        if(backLogItem.getDescription() == null || backLogItem.getDescription().equals("") || backLogItem.getPriority() == null || backLogItem.getPriority().equals("")
        || backLogItem.getEstimatedTime() == 0)
        {
            return null;
        }
        ProjectDB projectDB = socket.createBacklogItem(projectName, backLogItem);
        return projectDB.convertToProject();
    }

    public Sprint setUserForBacklog(String username, String id) {
        BackLogItemDB initialBacklogItem = socket.getBacklogItemById(id);

        if (initialBacklogItem.getWorksOnIt()==null)
        {
            SprintDB sprintDB = socket.setUserForBacklog(username, id);
            System.out.println(username + " is to be set as working in backlog item! Id: " + id);
            return sprintDB.convertToSprint();
        }
        else
            return
                null;


    }

    public Sprint updateActualTime(String id, String actualTime) {
        //Sounds weird, but DO NOT change the name of the method!!!S
        SprintDB sprint = socket.updateActualTime(id, actualTime);
        System.out.println(actualTime + " hours had been set for the backlog item, by id: " + id + "!");
        return sprint.convertToSprint();
    }

    public BackLogItem setSprintId(int backlogId, int sprintId) {
        BackLogItemDB bli = socket.setSprintId(backlogId, sprintId);
        System.out.println("Set sprint " + sprintId + " with backlog ID: " + backlogId);
        return bli.convertToBackLogItem();
    }

    public Sprint createSprint(String projectName, int[] backlogIdArray)    {
        int number=0;
        int[] sprintNumbers = socket.getSprintNumbers(projectName);

        if (sprintNumbers==null)
        {
            number=1;
            socket.createSprint(projectName,number);

        }
       else
        {
            number = sprintNumbers.length+1;

           socket.createSprint(projectName,number);
        }
            int sprintId = socket.getSprintId(projectName,number);
            for (int i=0; i<backlogIdArray.length;i++)
            {
                socket.setSprintId(backlogIdArray[i],sprintId);
            }
            SprintDB sprintDB =  socket.getSprint(projectName,number);
            return sprintDB.convertToSprint();
    }

    public UserStory addUserStory(String projectName, String description) {
        UserStory us = socket.addUserStory(projectName, description);
        System.out.println("New user story: " + description);
        return us;

    }

    public UserStory[] GetUserStoriesForProject(String pName) {
        UserStoryDB[] storiesDB = socket.getUserStories(pName);
        if (storiesDB != null) {
            UserStory[] stories = new UserStory[storiesDB.length];
            for (int i = 0; i < storiesDB.length; i++)
                stories[i] = storiesDB[i].convertToUserStory();
            return stories;
        } else
            return new UserStory[0];
    }

    public UserStory[] editUserStoryDescription(String projectName, String oldDescrition, String newDescriptino) {
        socket.editDescription(projectName, oldDescrition,newDescriptino);
        System.out.println("Edit user story description in project: " + projectName + " With new description: " + newDescriptino);
        return GetUserStoriesForProject(projectName);
    }

    public BackLogItem editDescriptionBacklogItem(String backlogId, String description)    {
        BackLogItemDB backLogItemDB = socket.editDescriptionBacklogItem(backlogId,description);
        return  backLogItemDB.convertToBackLogItem();

    }

    public UserStory editNumber(String projectName, String o, String n) {
        UserStory us = socket.editNumber(projectName, o, n);
        System.out.println("Edit number in project: " + projectName + "Old: " + o + " New: " + n);
        return us;
    }

    public String[] getAllUsernames()    {
        String[] usenames = socket.getAllUsernames();
        return  usenames;
    }

    public String[] getAllProjectNames()    {
        String[] projectNames = socket.getAllProjectNames();
        return  projectNames;
    }
}

File: Data/DatabaseModels/DataModel.cs
using Data.DataAccess;
using System;
using System.Collections.Generic;
using System.Text;

namespace Data.DatabaseModels
{
    public class DataModel
    {
        
        UserDataAccess userDataAccess = new UserDataAccess();
        ProjectWorkgroupDataAccess projectDataAccess = new ProjectWorkgroupDataAccess();
        InvitationDataAccess invitationDataAccess = new InvitationDataAccess();
        BacklogItemDataAccess backlogItemDataAccess = new BacklogItemDataAccess();
        UserStoryDataAccess userStoryDataAccess = new UserStoryDataAccess();
        SprintDataAccess sprintDataAccess = new SprintDataAccess();
        public User searchUser(string username)
        {
            Console.WriteLine("Searching for user in database");
            User user = userDataAccess.retrieveUserByName(username);
            return user;
        }
        public User addUser(User user)
        {
            Console.WriteLine("Adding new user to database");
            userDataAccess.addUser(user);
            var tempUser = userDataAccess.retrieveUserByName(user.username);
            return tempUser;
        }
        //creates a project in database and returns it
        public Project addNewProject(string username, string projectName)
        {
            Console.WriteLine("Adding new project to database");
            projectDataAccess.addProject(projectName, username);
            Project project = projectDataAccess.getProject(projectName);
            return project;
        }
        //given the username of the user return an array of projects that user is part of
        public Project[] getUserProjects(string username)
        {
            Console.WriteLine("Looking for projects by username");
            Project[] projects = projectDataAccess.getProjectsForUser(username);
            return projects;
        }
        //Invites user to project and gives back the User object fro which an invitation has been created.
        public User inviteUserToProject(string projectName, string username)
        {
            Console.WriteLine("Add user to a project");
            Console.WriteLine(projectName, username);
            User tempUser = null;
            tempUser = invitationDataAccess.AddInvitation(projectName, username);
            return tempUser;
        }
        //Deletes a user invitation and then gives back the User object for which an invitation has been deleted.
        public User deleteInvitation(string projectName, string username)
        {
            Console.WriteLine("Delete invitation for user" + username);
            Console.WriteLine(projectName, username);
            User tempUser = null;
            tempUser = invitationDataAccess.DeleteInvitation(projectName, username);
            return tempUser;
        }
        //Returns a string[] of Project names
        public string[] getInvitationsByUser(string InvitedUser)
        {
            Console.WriteLine("Fetch invitation for user " + InvitedUser);
            string[] stringArrayProjects = invitationDataAccess.GetInvitationsByUser(InvitedUser);
            return stringArrayProjects;
        }
        public Project AcceptInvitation(string projectName, string invitedUser)
        {
            Console.WriteLine("Accepts invitation for project" + projectName + "for user" + invitedUser);
            Project acceptedProjectInvitation = invitationDataAccess.AcceptInvitation(projectName, invitedUser);
            acceptedProjectInvitation = projectDataAccess.getProject(acceptedProjectInvitation.Name);
            return acceptedProjectInvitation;
        }
        public WorkGroup GetWorkGroup(string pName)
        {
            Console.WriteLine("Getting WorkGroup");
            return projectDataAccess.GetWorkGroup(pName);
        }
        public Project setScrumMaster(string username, string pName)
        {
            Console.WriteLine("Setting SM role in project" + pName + "for user" + username);
            projectDataAccess.SetScrumMaster(username, pName);
            return projectDataAccess.getProject(pName);
        }
        public Project updateScrumMaster(string username, string pName)
        {
            Console.WriteLine("Updating SM role in project" + pName + "for user" + username);
            projectDataAccess.UpdateExistingScrumMaster(username, pName);
            return projectDataAccess.getProject(pName);
        }
        public Project setProductOwner(string username, string pName)
        {
            Console.WriteLine("Setting PO role in project" + pName + "for user" + username);
            projectDataAccess.SetProductOwner(username, pName);
            return projectDataAccess.getProject(pName);
        }
        public Project updateProductOwner(string username, string pName)
        {
            Console.WriteLine("Updating PO role in project" + pName + "for user" + username);
            projectDataAccess.UpdateExistingProductOwner(username, pName);
            return projectDataAccess.getProject(pName);
        }
        public Project createBacklogItem(BackLogItem backlogItem)
        {
            Console.WriteLine("Creating backlog item");
            backlogItemDataAccess.CreateBacklogItem(backlogItem);
            Project project = projectDataAccess.assembleProject(backlogItem.Name);
            Console.WriteLine(project.BackLogItems.Count + " Im in DataModel****************************");
            return project;
        }
        public BackLogItem getBacklogItemById(int id)
        {
            return backlogItemDataAccess.getBacklogItemById(id);
        }
        public Sprint setUserForBacklog(int backlogId, string username)
        {
            Console.WriteLine("updateUsername");
            return backlogItemDataAccess.SetWorkingUser(backlogId, username);
        }
        public Sprint updateActualTime(int backlogId, int actualTime)
        {
            Console.WriteLine("updateActualTime");
            return backlogItemDataAccess.SetActualTime(backlogId, actualTime);
        }
        public BackLogItem setSprintId(int backlogId, int sprintId)
        {
            Console.WriteLine("setSprintId");
            return backlogItemDataAccess.SetSprintId(backlogId, sprintId);
        }
        public UserStory AddUserStory(UserStory story)
        {
            Console.WriteLine("AddUserStory");
            return userStoryDataAccess.AddUserStory(story);
        }
        public UserStory[] GetUserStoriesForProject(string projectName)
        {
            Console.WriteLine("GetUserStoriesForProject");
            return userStoryDataAccess.GetUserStoriesForProject(projectName);
        }
        public UserStory[] EditUserStoryDescription(string projectName, string oldDescription, string newDescription)
        {
            Console.WriteLine("EditUserStoryDescription");
            return userStoryDataAccess.EditDescription(projectName, oldDescription, newDescription);
        }
        public BackLogItem EditDescriptionBacklogItem(int backlogId, string description)
        {
            Console.WriteLine("EditDescriptionBacklogItem");
            return backlogItemDataAccess.EditDescription(backlogId, description);
        }
        public Sprint CreateSprint(string projectName, int number)
        {
            Console.WriteLine($"Add sprint {number} to project {projectName}");
            return sprintDataAccess.CreateSprint(projectName,number);
        }
        public Sprint[] GetSprintsForProject(string projectName)
        {
            Console.WriteLine($"Got sprints for project {projectName}");
            return sprintDataAccess.AssembleSprintsForProject(projectName);
        }
        public int getSprintId(string projectName, int sprintNumber)
        {
            Console.WriteLine($"Getting sprint id from {projectName}");
            return sprintDataAccess.GetSprintId(projectName, sprintNumber);
        }
        public int[] GetSprintNumbers(string projectName)
        {
            Console.WriteLine($"Getting sprint numbers for project {projectName}");
            return sprintDataAccess.GetSprintNumbers(projectName);
        }
        public Sprint GetSprint(string projectName,int sprintNumber)
        {
            return sprintDataAccess.GetSprint(projectName, sprintNumber);
        }
        public string[] GetAllUsernames()
        {
            Console.WriteLine("Getting all usernames");
            return userDataAccess.GetAllUsernames();
        }
        public string[] GetAllProjectNames()
        {
            Console.WriteLine("Getting all project names");
            return projectDataAccess.GetAllProjectNames();
        }
    }
}
